File: app/src/main/java/com/tpeapp/handler/ChatRepository.kt
package com.tpeapp.handler

import android.content.Context
import android.util.Log
import androidx.preference.PreferenceManager
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * ChatRepository — persists the conversation history in SharedPreferences and
 * sends messages to an OpenAI-compatible `/v1/chat/completions` endpoint.
 *
 * ### Dom-configurable prefs (writable via FCM)
 * | Key                          | Purpose                                         |
 * |------------------------------|-------------------------------------------------|
 * | [PREF_HANDLER_ENDPOINT]      | Base URL of the chat API, e.g. `https://api.openai.com` |
 * | [PREF_HANDLER_API_KEY]       | Bearer token / API key for the endpoint         |
 * | [PREF_HANDLER_SYSTEM_PROMPT] | System prompt that defines Handler's personality |
 * | [PREF_HANDLER_MODEL]         | Model name, e.g. `gpt-4o`                      |
 *
 * History is capped at [MAX_HISTORY] messages to avoid unbounded SharedPreferences growth.
 * The system prompt is *not* stored in the history list — it is prepended fresh on every
 * API call.
 */
object ChatRepository {

    private const val TAG = "ChatRepository"

    // ------------------------------------------------------------------
    //  SharedPreferences keys
    // ------------------------------------------------------------------

    const val PREF_HANDLER_ENDPOINT      = "handler_endpoint"
    const val PREF_HANDLER_API_KEY       = "handler_api_key"
    const val PREF_HANDLER_SYSTEM_PROMPT = "handler_system_prompt"
    const val PREF_HANDLER_MODEL         = "handler_model"
    private const val PREF_CHAT_HISTORY  = "handler_chat_history_json"

    // ------------------------------------------------------------------
    //  Defaults
    // ------------------------------------------------------------------

    private const val DEFAULT_ENDPOINT      = "https://api.openai.com"
    private const val DEFAULT_MODEL         = "gpt-4o"
    private const val DEFAULT_SYSTEM_PROMPT =
        "You are Handler, a strict but caring AI companion in a TPE (Total Power Exchange) " +
        "dynamic. You speak with authority and warmth. You hold the sub accountable to their " +
        "rules, offer guidance, and track their progress. You may use the word 'Handler' to " +
        "refer to yourself. Keep replies concise unless the sub needs detailed guidance."

    /** Maximum number of messages kept in history (oldest are pruned first). */
    private const val MAX_HISTORY = 100

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    // ------------------------------------------------------------------
    //  History persistence
    // ------------------------------------------------------------------

    fun getHistory(ctx: Context): List<ChatMessage> {
        val json = PreferenceManager.getDefaultSharedPreferences(ctx)
            .getString(PREF_CHAT_HISTORY, null) ?: return emptyList()
        return try {
            val arr = JSONArray(json)
            List(arr.length()) { i ->
                val o = arr.getJSONObject(i)
                ChatMessage(
                    id        = o.getString("id"),
                    role      = o.getString("role"),
                    content   = o.getString("content"),
                    timestamp = o.getLong("timestamp")
                )
            }
        } catch (e: Exception) {
            Log.w(TAG, "Failed to parse chat history", e)
            emptyList()
        }
    }

    fun addMessage(ctx: Context, message: ChatMessage): List<ChatMessage> {
        val list = getHistory(ctx).toMutableList()
        list.add(message)
        if (list.size > MAX_HISTORY) {
            list.subList(0, list.size - MAX_HISTORY).clear()
        }
        saveHistory(ctx, list)
        return list.toList()
    }

    fun clearHistory(ctx: Context) {
        PreferenceManager.getDefaultSharedPreferences(ctx).edit()
            .remove(PREF_CHAT_HISTORY).apply()
    }

    // ------------------------------------------------------------------
    //  API call — blocking; call on a background thread / coroutine
    // ------------------------------------------------------------------

    /**
     * Sends the user's [userMessage] (plus history context) to the configured
     * OpenAI-compatible endpoint and returns the assistant's reply text.
     *
     * Throws an [Exception] on network failure or non-200 response.
     */
    fun sendMessage(ctx: Context, userMessage: String): String {
        val prefs    = PreferenceManager.getDefaultSharedPreferences(ctx)
        val endpoint = prefs.getString(PREF_HANDLER_ENDPOINT, DEFAULT_ENDPOINT)
            ?.trimEnd('/') ?: DEFAULT_ENDPOINT
        val apiKey   = prefs.getString(PREF_HANDLER_API_KEY, null)?.takeIf { it.isNotBlank() }
        val model    = prefs.getString(PREF_HANDLER_MODEL, DEFAULT_MODEL) ?: DEFAULT_MODEL
        val systemPrompt = prefs.getString(PREF_HANDLER_SYSTEM_PROMPT, DEFAULT_SYSTEM_PROMPT)
            ?.takeIf { it.isNotBlank() } ?: DEFAULT_SYSTEM_PROMPT

        // Build the messages array: [system] + recent history + [new user message]
        val messages = JSONArray()
        messages.put(JSONObject().apply { put("role", "system"); put("content", systemPrompt) })

        // Include only the last 20 turns to stay within context window limits
        val history = getHistory(ctx).takeLast(20)
        history.forEach { msg ->
            messages.put(JSONObject().apply { put("role", msg.role); put("content", msg.content) })
        }
        messages.put(JSONObject().apply { put("role", "user"); put("content", userMessage) })

        val requestBody = JSONObject().apply {
            put("model",    model)
            put("messages", messages)
        }.toString().toRequestBody("application/json".toMediaType())

        val reqBuilder = Request.Builder()
            .url("$endpoint/v1/chat/completions")
            .post(requestBody)
            .addHeader("Content-Type", "application/json")
        if (apiKey != null) reqBuilder.addHeader("Authorization", "Bearer $apiKey")

        httpClient.newCall(reqBuilder.build()).execute().use { response ->
            val body = response.body?.string() ?: throw Exception("Empty response body")
            if (!response.isSuccessful) {
                Log.w(TAG, "API error ${response.code}: $body")
                throw Exception("Handler API error ${response.code}")
            }
            val json = JSONObject(body)
            return json
                .getJSONArray("choices")
                .getJSONObject(0)
                .getJSONObject("message")
                .getString("content")
        }
    }

    // ------------------------------------------------------------------
    //  Setters (called by FCM handlers)
    // ------------------------------------------------------------------

    fun setEndpoint(ctx: Context, endpoint: String) {
        PreferenceManager.getDefaultSharedPreferences(ctx).edit()
            .putString(PREF_HANDLER_ENDPOINT, endpoint).apply()
    }

    fun setApiKey(ctx: Context, key: String) {
        PreferenceManager.getDefaultSharedPreferences(ctx).edit()
            .putString(PREF_HANDLER_API_KEY, key).apply()
    }

    fun setSystemPrompt(ctx: Context, prompt: String) {
        PreferenceManager.getDefaultSharedPreferences(ctx).edit()
            .putString(PREF_HANDLER_SYSTEM_PROMPT, prompt).apply()
    }

    fun setModel(ctx: Context, model: String) {
        PreferenceManager.getDefaultSharedPreferences(ctx).edit()
            .putString(PREF_HANDLER_MODEL, model).apply()
    }

    // ------------------------------------------------------------------
    //  Internal
    // ------------------------------------------------------------------

    private fun saveHistory(ctx: Context, list: List<ChatMessage>) {
        val arr = JSONArray()
        list.forEach { m ->
            arr.put(JSONObject().apply {
                put("id", m.id); put("role", m.role)
                put("content", m.content); put("timestamp", m.timestamp)
            })
        }
        PreferenceManager.getDefaultSharedPreferences(ctx).edit()
            .putString(PREF_CHAT_HISTORY, arr.toString()).apply()
    }

    /** Convenience factory for a new user message. */
    fun newUserMessage(text: String) = ChatMessage(
        id        = UUID.randomUUID().toString(),
        role      = "user",
        content   = text,
        timestamp = System.currentTimeMillis()
    )

    /** Convenience factory for a new assistant message. */
    fun newAssistantMessage(text: String) = ChatMessage(
        id        = UUID.randomUUID().toString(),
        role      = "assistant",
        content   = text,
        timestamp = System.currentTimeMillis()
    )
}

File: flutter_app/lib/services/chat_repository.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:uuid/uuid.dart';

import '../models/chat_message.dart';

/// Dart equivalent of [com.tpeapp.handler.ChatRepository].
///
/// Persists chat history in SharedPreferences and calls an
/// OpenAI-compatible `/v1/chat/completions` endpoint.
class ChatRepository extends ChangeNotifier {
  ChatRepository(this._prefs) {
    _load();
  }

  static const endpointKey     = _endpointKey;
  static const apiKeyKey       = _apiKeyKey;
  static const modelKey        = _modelKey;
  static const systemPromptKey = _systemPromptKey;

  static const _historyKey      = 'handler_chat_history_json';
  static const _endpointKey     = 'handler_endpoint';
  static const _apiKeyKey       = 'handler_api_key';
  static const _systemPromptKey = 'handler_system_prompt';
  static const _modelKey        = 'handler_model';

  static const _defaultEndpoint = 'https://api.openai.com';
  static const _defaultModel    = 'gpt-4o';
  static const _defaultPrompt   =
      'You are Handler, a strict but caring AI companion in a TPE (Total Power Exchange) '
      'dynamic. You speak with authority and warmth. You hold the sub accountable to their '
      'rules, offer guidance, and track their progress. You may use the word \'Handler\' to '
      'refer to yourself. Keep replies concise unless the sub needs detailed guidance.';
  static const _maxHistory = 100;

  final SharedPreferences _prefs;
  List<ChatMessage> _history = [];

  List<ChatMessage> get history => List.unmodifiable(_history);

  String get endpoint =>
      _prefs.getString(_endpointKey)?.trimRight() ?? _defaultEndpoint;
  String? get apiKey =>
      _prefs.getString(_apiKeyKey)?.let((k) => k.isNotEmpty ? k : null);
  String get model => _prefs.getString(_modelKey) ?? _defaultModel;
  String get systemPrompt =>
      _prefs.getString(_systemPromptKey)?.let((p) => p.isNotEmpty ? p : null) ??
      _defaultPrompt;

  void _load() {
    final json = _prefs.getString(_historyKey);
    if (json == null) return;
    try {
      final list = jsonDecode(json) as List<dynamic>;
      _history =
          list.map((e) => ChatMessage.fromJson(e as Map<String, dynamic>)).toList();
    } catch (_) {
      _history = [];
    }
  }

  Future<void> _save() async {
    final encoded =
        jsonEncode(_history.map((m) => m.toJson()).toList());
    await _prefs.setString(_historyKey, encoded);
  }

  Future<ChatMessage> addUserMessage(String text) async {
    const uuid = Uuid();
    final msg = ChatMessage(
      id: uuid.v4(),
      role: 'user',
      content: text,
      timestamp: DateTime.now().millisecondsSinceEpoch,
    );
    _history.add(msg);
    if (_history.length > _maxHistory) {
      _history.removeRange(0, _history.length - _maxHistory);
    }
    await _save();
    notifyListeners();
    return msg;
  }

  Future<ChatMessage> addAssistantMessage(String text) async {
    const uuid = Uuid();
    final msg = ChatMessage(
      id: uuid.v4(),
      role: 'assistant',
      content: text,
      timestamp: DateTime.now().millisecondsSinceEpoch,
    );
    _history.add(msg);
    if (_history.length > _maxHistory) {
      _history.removeRange(0, _history.length - _maxHistory);
    }
    await _save();
    notifyListeners();
    return msg;
  }

  Future<void> clearHistory() async {
    _history.clear();
    await _prefs.remove(_historyKey);
    notifyListeners();
  }

  /// Sends [userText] to the configured OpenAI-compatible endpoint and returns
  /// the assistant's reply.  Throws on network failure or non-200 response.
  Future<String> sendMessage(String userText) async {
    final baseUrl = endpoint.endsWith('/') ? endpoint.dropLast(1) : endpoint;

    final recent = _history.takeLast(20).toList();

    final messages = [
      {'role': 'system', 'content': systemPrompt},
      ...recent.map((m) => {'role': m.role, 'content': m.content}),
      {'role': 'user', 'content': userText},
    ];

    final headers = {
      'Content-Type': 'application/json',
      if (apiKey != null) 'Authorization': 'Bearer $apiKey',
    };

    final body = jsonEncode({'model': model, 'messages': messages});

    final response = await http
        .post(Uri.parse('$baseUrl/v1/chat/completions'),
            headers: headers, body: body)
        .timeout(const Duration(seconds: 60));

    if (response.statusCode != 200) {
      throw Exception('Handler API error ${response.statusCode}: ${response.body}');
    }

    final decoded = jsonDecode(response.body) as Map<String, dynamic>;
    return (decoded['choices'] as List<dynamic>)[0]['message']['content']
        as String;
  }

  // ── Settings setters (called when FCM updates arrive) ────────────────

  Future<void> setEndpoint(String url) async {
    await _prefs.setString(_endpointKey, url);
  }

  Future<void> setApiKey(String key) async {
    await _prefs.setString(_apiKeyKey, key);
  }

  Future<void> setSystemPrompt(String prompt) async {
    await _prefs.setString(_systemPromptKey, prompt);
  }

  Future<void> setModel(String m) async {
    await _prefs.setString(_modelKey, m);
  }
}

// ── Private extension helpers ──────────────────────────────────────────

extension _StringX on String {
  T let<T>(T Function(String) block) => block(this);
  String dropLast(int n) => substring(0, length - n);
}

extension<T> on List<T> {
  List<T> takeLast(int n) => length <= n ? this : sublist(length - n);
}
